### PR TITLE
Android GameActivity example

### LINKED ISSUE #12026: Android example based on GameActivity
## What problem does this solve or what need does it fill?

1) The mobile example (bevy 0.13) doesn't work on my devices <= Android 8.1. When I did get it to work (bevy 0.12) the button was offset from the place where you touch.

2) GameActivity works better, is the current recommended approach and a better match for Bevy. NativeActivity is legacy, crashes on Android <= 5.1 and doesn't have a surface needed for the onscreen keyboard.

3) Cargo apk is deprecated and too limited. GameActivity is compiled separately using a normal Android project, which is what any serious game developer would need anyway. You can keep the Android project in a separate folder from the rust project; the rust part is compiled using cargo ndk.

4) wgpu currently does not work on Android without Vulkan if the backend is not set to GL.

5) The contents of the example could be improved; the zooming is weird and does not use the touch drag in a sensible way.

## What solution would you like?

I'm just reporting and not saying this is a priority, but bevy is going to need a GameActivity example since that is the appropriate way to go for a game engine.
It could be based on the agdk examples from rust-android-examples, it's not difficult.
Ideally it would demonstrate use of the keyboard as well (as abstracted by winit, but you need to pass around AndroidApp in order to show the keyboard).

### PR DESCRIPTION
**Pull request by Cyannide**
# Objective

- Updates the Android example to use AGDK GameActivity.
- Also replaced the cargo-apk based build with Gradle and cargo-ndk.
- Fixes #12026.

## Solution

- Replaced the winit feature [android-native-activity] with [android-game-activity] and added a Gradle based build system to build the native libraries and package everything into an APK for installation.

---

## Changelog

- Removed the cargo-apk build capability.
- Changed winit to use android-game-activity feature.
- Added a Gradle based build to product the native libraries and final APK.

## Migration Guide

- Building/running using `cargo apk build/run -p bevy_mobile_example` is no longer possible.
- Move into the examples/mobile directory and run `./gradlew build`, `./gradlew installDebug`, `adb shell am start -n org.bevyengine.example/.MainActivity`

### CODE REVIEW

**Comment by MalekiRe** [comment]
Gradle is not good! Should use xbuild instead as a replacement for cargo-apk

**Comment by MalekiRe** [comment]
Gradle can be an option but shouldn't be the only one

**Comment by Cyannide** [comment]
> Gradle can be an option but shouldn't be the only one

I'll have a look at xbuild again and add it as another build method.

The main goal for me was to have a working GameActivity based example.

**Comment by mockersf** [comment]
> Gradle is not good! Should use xbuild instead as a replacement for cargo-apk

Could you explain why? Gradle is the standard tool for building things on Android

**Comment by Cyannide** [comment]
> > Gradle is not good! Should use xbuild instead as a replacement for cargo-apk
> 
> Could you explain why? Gradle is the standard tool for building things on Android

My thoughts exactly and after looking at xbuild, it uses Gradle as part of the build process.

Currently xbuild has no support for GameActivity which was the entire purpose of this PR, so without some modifications to xbuild, it isn't an option at the moment.

**Comment by extrawurst** [comment]
If this rules out using xbuild it would be great if you could add documentation on how to make this build an aab though cause I think people use xbuild for aab at the moment

**Comment by Cyannide** [comment]
I'm still looking into xbuild as well because it would be more convenient than having to manually set up the Gradle build.

**Comment by mockersf** [comment]
Could you change your PR to be based on the `main` branch instead? I tried to change in GitHub, but that causes a lot of conflicts

File: examples/mobile/Cargo.toml [hunk 1]
```diff
@@ -8,26 +8,10 @@ license = "MIT OR Apache-2.0"
 
 [lib]
 name = "bevy_mobile_example"
-crate-type = ["staticlib", "cdylib"]
```
Comment (mockersf) [inline examples/mobile/Cargo.toml:11]: this is needed for iOS

Comment (Cyannide) [reply examples/mobile/Cargo.toml:11]: Will fix it right  now. And I'll change to `main` branch. I was actually curious about what I should base it on.

**Comment by Salzian** [comment]
Any specific reason this is not written in Kotlin? Kotlin is Android's recommended programming language. The same goes for the gradle sccipts themselves. Kotlin DSL is the recommended language of choice here.

Other than that, I apprechiate the effort. This PR might make it possible to introduce third-party libraries like FMOD more easily. I'm currently trying it out on a Quest 3 device.

File: examples/mobile/app/src/main/java/org/bevyengine/examples/MainActivity.java
Comment (Salzian) [inline examples/mobile/app/src/main/java/org/bevyengine/examples/MainActivity.java:1]: Folder name should align with package name. Replace `examples` with `example`.

File: examples/mobile/build.gradle [hunk 2]
```diff
@@ -0,0 +1,25 @@
+// Top-level build file where you can add configuration options common to all sub-projects/modules.
+plugins {
+	id 'com.android.application' version '8.0.0' apply false
+	id 'com.android.library' version '8.0.0' apply false
+}
+
+task clean(type: Delete) {
+    delete rootProject.buildDir
+}
```
Comment (Salzian) [inline examples/mobile/build.gradle:9]: I don't understand this task. Isn't this already the default behaviour of the clean task?

File: examples/mobile/build.gradle [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+            include "sounds/Windless Slopes.ogg"
+            into layout.projectDirectory.dir("app/src/main/assets")
+        }
+    }
+}
+
+task ndkBuild(type: Exec) {
+    workingDir layout.projectDirectory
+    commandLine 'cargo', 'ndk', '-t', 'arm64-v8a', '-t', 'armeabi-v7a', '-o', 'app/src/main/jniLibs/', 'build'
+}
```
Comment (Salzian) [inline examples/mobile/build.gradle:25]: You should migrate these tasks to use `tasks.register`. More info: https://docs.gradle.org/current/userguide/task_configuration_avoidance.html

File: examples/mobile/app/build.gradle [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+    // To use the Games Activity library
+    //implementation("androidx.games:games-activity:3.0.2")		// 3.0.2 results in an UnsatisfiedLinkError due to the change in onTouchEventNative method signature
+    implementation("androidx.games:games-activity:2.0.2")
+
+    // To use the Games Controller Library
+    //implementation "androidx.games:games-controller:1.1.0"
+
+    // To use the Games Text Input Library
+    //implementation "androidx.games:games-text-input:1.1.0"
```
Comment (Salzian) [inline examples/mobile/app/build.gradle:76]: SOme of these are outdated. Now, it doesn't make much sense to keep chasing the latest versions for example code, but I would recommend adding a comment that points this out.

**Review by Salzian — COMMENTED** [review]
Some minor things, but overall, I'm very happy to see this code. cargo apk and xbuild were alsways weird to work with. They are good at covering the basics, but the moment you need to do something android specific, like loading external libraries, they stop being very useful. I definitely tly prefer a gradle project.

File: examples/mobile/app/src/main/java/org/bevyengine/examples/MainActivity.java
Comment (Cyannide) [reply examples/mobile/app/src/main/java/org/bevyengine/examples/MainActivity.java:1]: Ahh yes, forgot to move it when I removed the 's' in other places.

File: examples/mobile/build.gradle (hunk 2 above)
Comment (Cyannide) [reply examples/mobile/build.gradle]: I added the other required cleanup into the task.

File: assets/android-res/mipmap-mdpi/ic_launcher.png
Comment (Salzian) [inline assets/android-res/mipmap-mdpi/ic_launcher.png:1]: THis file got downsized by a lot. Is this expeceted?

**Review by Salzian — COMMENTED** [review]
Technically, this looks good to me. I will give it another try on my Quest later with the latest changes. 

I'm not sure if we should completely abandon the cargo-apk/xbuild approach (cargo-apk is deprecated, so probably stick to xbuild for now). xbuild is easy to get things started with and apparently also provides easy cross-compilation between Android and iOS.

I would still like to see this code merged, maybe as a separate example (something like `android-gradle`), for people, that want to have full control over their Android build. I personally will definitely start using this example as I work a lot with Kotlin and Gradle at work already.

But in the end, you need a review from a bevy person anyway, they have the final say. I'm just sharing my opinion with the bit of knowledge I have on JVM stuff.

File: assets/android-res/mipmap-mdpi/ic_launcher.png
Comment (Cyannide) [reply assets/android-res/mipmap-mdpi/ic_launcher.png:1]: Yeah, the original image was 256x256 pixels but the mdpi images should be 48x48.

**Comment by Cyannide** [comment]
I have successfully built and ran the example using xbuild but it is non-trivial and requires several manual steps to get everything in place. (copying libc++_shared.so, assets, resources)
If anyone is interested in pursuing the xbuild option, I'd be happy to share the details on getting it running, but without adding more flexibility to the Gradle portion of xbuild, it is quite cumbersome.

**Comment by mockersf** [comment]
#12095 has been merged

**Pull request closed without merging**